src/find_sort_location.rs

```diff
@@ -0,0 +1,142 @@
+#![allow(dead_code)]
+
+use futures::future::join_all;
+
+use crate::{
+    field::Field,
+    helpers::ring::Ring,
+    replicated_secret_sharing::ReplicatedSecretSharing,
+    securemul::{ProtocolContext, SecureMul},
+};
+
+pub struct SecureSort<'a, R, F> {
+    context: &'a ProtocolContext<'a, R>,
+    share_of_one: ReplicatedSecretSharing<F>,
+}
+
+impl<F: Field, R: Ring> SecureSort<'_, R, F> {
+    pub async fn execute(
+        self,
+        shares: &[ReplicatedSecretSharing<F>],
+    ) -> Vec<ReplicatedSecretSharing<F>> {
+        let gen_permutation_share = |mult_output: Vec<ReplicatedSecretSharing<F>>| {
+            let len = mult_output.len() / 2;
+
+            // Fold the result
+            let mut permutation_share = Vec::new();
+            (0..len).for_each(|j| {
+                permutation_share.push(mult_output[j] + mult_output[j + len]);
+            });
+            permutation_share
+        };
+        gen_permutation_share(self.secure_multiply(self.prepare_mult_inputs(shares)).await)
+    }
+
+    fn prepare_mult_inputs(
+        &self,
+        shares: &[ReplicatedSecretSharing<F>],
+    ) -> Vec<(ReplicatedSecretSharing<F>, ReplicatedSecretSharing<F>)> {
+        // Step1: Calculate x and 1 - x
```

Comment:
there's a different way you can do this function, using `scan`:

```rust
    shares
        .iter()
        .map(|x| self.share_of_one - *x)
        .chain(shares.iter())
        .scan(
            ReplicatedSecretSharing::<F>::new(F::ZERO, F::ZERO),
            |sum, n| {
                *sum += n;
                Some((n, *sum))
            },
        )
        .collect()
```

Reply:
This is really cool! Thanks for the suggestion!